### ISSUE TITLE
Specify the `grafanaplugin` package instance in CUE and adapt `pfs.Tree` accordingly

### ISSUE THREAD

**Issue by sdboyer** [grafana/grafana#59002]
We need a specification for the set of `.cue` files in package `grafanaplugin` that exist next to a `plugin.json`.

This is the same function served by `kindsys.#CoreStructured` to specify what's declared in `kinds/structured/*/*.cue`, and `kindsys.#Raw` for `kinds/raw/*/*.cue`.

With the core kinds, we take advantage of a filesystem pattern that allows there to be one kind declaration per directory. With plugins, we have to jam it all into a single directory and package - what CUE calls an "[instance](https://cuelang.org/docs/concepts/packages/#instances)" - for simplicity and clarity of both declarations for plugin authors and our loader logic.

Thus, we need to create a CUE definition that specifies the shape of that whole `grafanaplugin` instance within a plugin. There are a few key considerations, as i see it:

* There must be space to declare as many custom kinds as a plugin wants
* Some constraint must be enforced between a plugin's id, and the namespace of all declared kinds - this is an item tracked in #59001 

```cue
import "github.com/grafana/thema"

// GrafanaPluginInstance specifies what plugins may declare in .cue files within
// the `grafanaplugin` package, at the plugin root adjacent to plugin.json.
GrafanaPluginInstance: {
	// for now it's probably easiest to require the plugin id be re-declared here,
	// then we can check it's == to what's in plugin.json from Go code
	id: string & strings.MinRunes(1)

	// The composable kinds declared by a plugin
	composableKinds: #ComposableSlots
	// The custom kinds declared by a plugin
	customKinds: [string]: kindsys.#CustomStructured

	// it's almost certainly better to just change the way slots are declared to
	// avoid the need for this field comprehension with dynamic keys
	#ComposableSlots: {for slotname, meta in kindsys.slots {
		"\(slotname)"?: kindsys.#Composable & {
			slot: slotname
			lineage: { joinSchema: meta }
		}
	}}

	// Eventually it may be good to do something like the following, in order to get
	// the plugin.json and grafanaplugin CUE instance all loaded into the same space

	// The Go pfs pkg needs to either cue.Value.FillPath on this with plugin.json
	// contents, or inject the plugin.json contents into cuelang.org/go/build.Instance.
	// latter is probably preferable for thread safety
	_pluginst: _

	// The input plugin.json, translated to the latest schema version, ignoring lacunas.
	let pjson = (thema.#Translate & {
		to:    (thema.#Latest & {lin: plugindef}).out
		linst: (thema.#ValidateAny & {
			// This pseudofunc is the same as Lineage.ValidateAny in Go - given some data,
			// find the schema version it's valid against, if any.
			lin:  plugindef
			inst: _pluginst
		}).out
	}).out.linst.inst

}

```

**Comment by mckn** [comment]
I really like this idea! We should probably also add a way to define plugin internal types that still should be generated to backend/frontend.

**Comment by sdboyer** [comment]
and where i landed on fundamental constraints is:

* Every `*.cue` file that is sibling to `plugin.json` MUST be parsed
  * There MAY be zero files matching the `*.cue` pattern
* ONLY parsed `*.cue` files with `package grafanaplugin` are loaded/built into a `cue.Value` that we will refer to as "the package instance". CUE files in any package other than `grafanaplugin` MAY exist, but MUST be discarded by this loader
  * There MAY be zero `*.cue` files with `package grafanaplugin` - this is treated as equivalent to there being no `*.cue` files at all
* The package instance MAY contain `import` statements. but any imports MUST be on an explicit allowlist ([currently these](https://github.com/grafana/grafana/blob/ee8f292c6a536a7b5cf981abbc10cbcbe65e3fc9/pkg/plugins/pfs/pfs.go#L22-L31))
  * To be clear, this also means the package instance MAY NOT import `.cue` files even from elsewhere within the plugin’s own tree, even in sibling `.cue` files from a different package
* The package instance MUST be a valid instance of what i sketched out in OP as `GrafanaPluginInstance`

**Comment by sdboyer** [comment]
cc @IfSentient

**Comment by mckn** [comment]
I think most of what you have written here is good. Please see my comments for each section for more details.

> The OP articulates what an instance should look like, but it doesn't explain the key constraints about what's expected/allowed on disk - all the prerequisites that must be met before we can talk about what's actually been declared by a plugin. It's worth writing those down, too:
> 
> * Every `*.cue` file that is sibling to `plugin.json` MUST be parsed

If you would have a plugin with a couple of schemas and migrations defined for them, you will probably end up in a situation where you either have really large files or lots of files. In this case you probably want to have multiple files/folders to structure your schemas.  

If we would follow the conventions defined here: https://cuelang.org/docs/concepts/packages/ we would give the developer the possibility to structure their code in a easy and maintainable way. We could also reference the cuelang documentation instead of providing all documentation our self.

>   * There MAY be zero files matching the `*.cue` pattern

OK!

> * ONLY parsed `*.cue` files with `package grafanaplugin` are loaded/built into a `cue.Value` that we will refer to as "the package instance".

Would it be possible to specify a convention where you specify the plugin package (according to the description of this issue) in your plugin.cue and it needs to have `package grafanaplugin` or maybe `package plugin` to make it shorter. We only evaluate the plugin package but the plugin author would be able to import schemas within the same module as the package plugin is specified in.

```bash
    .
   ├── cue.mod 
   │   ├── module.cue 
   │   └── schemas 
   │       ├── plugin.cue
   │       ├── entities
   │       │   ├── incident.cue
   │       │   └── ... 
   │       └── ... 
   ├── plugin.json    #we will move this to the root from the src
   ├── pkg
   ├── src
   └── ...
```

>   * CUE files in any package other than `grafanaplugin` MAY exist, but MUST be discarded by a loader

Can you expand a bit more on this?

>   * There MAY be zero `*.cue` files with `package grafanaplugin` - this is treated as equivalent to there being no `*.cue` files at all

OK!

> * The package instance MAY contain `import` statements. but any imports MUST be on an explicit allowlist ([currently these](https://github.com/grafana/grafana/blob/ee8f292c6a536a7b5cf981abbc10cbcbe65e3fc9/pkg/plugins/pfs/pfs.go#L22-L31))
>   

Can you expand a bit more on this?

>   * To be clear, this also means the package instance MAY NOT import `.cue` files even from elsewhere within the plugin’s own tree, even in sibling `.cue` files from a different package

Can you expand a bit more on this?

> * The package instance MUST be a valid instance of what i sketched out in OP as `GrafanaPluginInstance`

OK! Could we remove the `id`? We already have that one in the plugin.json and it would be good to try to keep it as a single source of truth for the plugin id.

**Comment by sdboyer** [comment]
lots of excellent questions! delighted to see you thinking so expansively about this

> If you would have a plugin with a couple of schemas and migrations defined for them, you will probably end up in a situation where you either have really large files or lots of files. In this case you probably want to have multiple files/folders to structure your schemas.

I certainly imagine this will be the case - and not just here, but for thema in general. The best answer i currently have is that we think of the current (and planned - grafana/thema#82) form for declaring a lineage to be the canonical normal form which is an OK-but-not-great UX, and that we'll eventually allow the introduction of some forms that make it easier to manage the declarations at scale.

My main reason for wanting to defer on additional forms is because i am concerned about the complexity cost that additional forms _may_ have on writing thema invariant checkers (most of which i still have yet to write). Until those are written and i can concretely evaluate the cost of adding support for more forms, i'm loathe to commit to any expanded scope, as it will delay delivery of a Thema MVP for a day 2/scaling use case.

> If we would follow the conventions defined here: https://cuelang.org/docs/concepts/packages/ we would give the developer the possibility to structure their code in a easy and maintainable way. We could also reference the cuelang documentation instead of providing all documentation our self.

I spun in circles for quite a while on how to apply CUE's native package concepts to what we're doing. Eventually, i reached the conclusion that those concepts were primarily designed with as-code repos in mind. At the level of schema/kind declaration in plugins, what we're doing just looks different. For example, there's no need for us to think about [root/medial/leaf](https://cuelang.org/docs/concepts/packages/#instances) directories in the context of a plugin, because our use pattern is not "iterate over and load a set of leaf directories in order to produce concrete objects." For the use case of loading kind declarations, we want to load and validate all of them at once, and none of them are expected to be concrete.

Rather, we should look at what's declared in a plugin's `grafanaplugin` package as an element within that strategy - folks could plausibly `import "github.com/grafana/grafana/public/app/plugins/panel/timeseries:grafanaplugin` from a root directory in their as-code repo.

> or maybe `package plugin` to make it shorter.

The main reason for `package grafanaplugin` to be long, weird, and specific, in order to minimize the possibility of any filesystem arrangement that unintentionally causes extra .cue files to be picked up by the hierarchical loading rules discussed above.

> We only evaluate the plugin package but the plugin author would be able to import schemas within the same module as the package plugin is specified in.

First, `cue.mod` is generally more like a `.git` directory - it marks the root of a the module tree, but your .cue files generally aren't supposed to go in it.

Using a tree to illustrate seems helpful, i'll do that. As i think of it, we start with literally just `plugin.json`:

```bash
.
└── plugin.json
```

In external plugins, at least, there'll likely need to exist a `cue.mod`:

```bash
.
├── cue.mod
│   └── module.cue
└── plugin.json
```

But nothing else is in that directory.

Today, we have models.cue at the root:

```bash
.
├── cue.mod
│   └── module.cue
├── models.cue # in package grafanaplugin, loaded
└── plugin.json
```

And that gets loaded by logic called from `pfs.ParsePluginFS`. One shift i'm suggesting here is that we should load _any_ file sibling to `plugin.json` in the appropriate package:

```bash
.
├── cue.mod
│   └── module.cue
├── grafanaplugin_file.cue # also in package grafanaplugin, loaded
├── models.cue
└── plugin.json
```

But, to expand on this as requested

> > CUE files in any package other than grafanaplugin MAY exist, but MUST be discarded by a loader
> Can you expand a bit more on this?

i'm saying that it's valid for `other_pkg.cue` to exist, but `pfs.ParsePluginFS` must ignore it:

```bash
.
├── cue.mod
│   └── module.cue
├── grafanaplugin_file.cue
├── models.cue
├── other_pkg.cue # in package "other", discarded
└── plugin.json
```

Similarly, if there are subdirectories containing files, the loader will not automatically see them:

```bash
.
├── cue.mod
│   └── module.cue
├── grafanaplugin_file.cue
├── models.cue
├── other_pkg.cue
├── plugin.json
└── schemas
    └── grafanaplugin_file.cue # not sibling to plugin.json, not loaded
```

In the future, we might allow `schemas/grafanaplugin_file.cue` or `other_pkg.cue` to be pulled in by `pfs.ParsePluginFS` - but only indirectly, by having an `import` statement in either `models.cue` or `grafanaplugin_file.cue` that pulls the contents of those files into the instance that `pfs.ParsePluginFS` is actually trying to load.

However, for now:

> > The package instance MAY contain import statements. but any imports MUST be on an explicit allowlist ([currently these](https://github.com/grafana/grafana/blob/ee8f292c6a536a7b5cf981abbc10cbcbe65e3fc9/pkg/plugins/pfs/pfs.go#L22-L31))
> Can you expand a bit more on this?

We should not allow this, for two reasons:

**Thema correctness** - basically, Thema's invariants can only _ever_ be meaningful if the text it's checking over for backwards compatibility, and translatability is immutable. Allowing imports of external elements is the easiest way to blow up that guarantee. Disallowing imports is neither necessary nor sufficient to address this issue, but it is a quick and easy way of doing most of what we need.

**MVP Implementation complexity** - in order to generate thema bindings (the equivalent of [this file](https://github.com/grafana/grafana/blob/49f91721ec4d76076c2de4eb2321e3039c60571a/pkg/plugins/plugindef/plugindef_bindings_gen.go), but for use in a plugin), there has to be an `embed.FS` that contains the `cue.mod` and any relevant files to load, _transitively including any imported files._ Correctly injecting the allowlisted packages is the job of functions like [this one](https://github.com/grafana/grafana/blob/49f91721ec4d76076c2de4eb2321e3039c60571a/pkg/cuectx/ctx.go#L145). But the plugin-declared bits still need to be rolled into an `fs.FS` for passing to `pfs.ParsePluginFS` - that's what's in our `embed.FS`. With the constraints i propose, this will _always_ be sufficient for that loader:

```go
//go:embed cue.mod *.cue
var CueFS embed.FS
```

But if we allow imports, even plugin module-internal ones, we'd need more adaptive logic to decide what goes in the embed. Getting these loaders to work cleanly is deceptively challenging, and could easily delay initial delivery by weeks. Given that this is all primarily oriented towards a day 2+ scaling problem, i'm suggesting we see how far we can get with these more restrictive rules.

Now, in that `embed.FS`, the `other_file.cue` would have been included. So why disallow that in the name of MVP?

> > To be clear, this also means the package instance MAY NOT import .cue files even from elsewhere within the plugin’s own tree, even in sibling .cue files from a different package
> Can you expand a bit more on this?

Today, [this is the check](https://github.com/grafana/grafana/blob/49f91721ec4d76076c2de4eb2321e3039c60571a/pkg/plugins/pfs/pfs.go#L209-L211) to decide if an import is allowed. It's a trivial map lookup.

To allow imports of other sibling `.cue` files from a different package, that check ceases to be just a static lookup, and becomes dynamically dependent on state within the plugin itself. We have to know whether a `cue.mod` exists, what its mod path is, what the path relation is between `cue.mod` and `plugin.json`, whether the import statement is pointing to the same path as in that relation.

And, since `pfs.ParsePluginFS` is dependent on an `fs.FS`, not the actual filesystem, there's always the possibility that what's handed to it is a misleadingly incomplete picture of what's on the filesystem - _deeply_ confusing for users ("why is it saying i don't have a `cue.mod`? it's right there!?!?!")

These aren't unsolvable problems, and it's probably worth tackling them eventually. I just see the costs of answering them, believe we can safely and sanely defer on them for MVP, and see only marginal benefit to answering them before we're actually feeling this anticipated friction.

> > The package instance MUST be a valid instance of what i sketched out in OP as GrafanaPluginInstance
> OK! Could we remove the id? We already have that one in the plugin.json and it would be good to try to keep it as a single source of truth for the plugin id.

Sorry, i explained this poorly in OP.

Yes, `id` is canonically written in `plugin.json`, and i'm not proposing a change. `id` should be included in `GrafanaPluginInstance` IFF:

1. There is some constraining relation we want to enforce between the plugin's `id` and the names for custom and composable kinds
2. We find it easy to enforce in `pfs.ParsePluginFS` that it == what's in `plugin.json`, so we're not undermining SSoT

Item 2 should be trivial. The big question is about item 1, which is why i made it its own item in [the epic tasklist](https://github.com/grafana/grafana/issues/59001).

**Comment by mckn** [comment]
> My main reason for wanting to defer on additional forms is because i am concerned about the complexity cost that additional forms _may_ have on writing thema invariant checkers (most of which i still have yet to write). Until those are written and i can concretely evaluate the cost of adding support for more forms, i'm loathe to commit to any expanded scope, as it will delay delivery of a Thema MVP for a day 2/scaling use case.

Narrowing the scope sounds good as long as we don't end up in a situation where we can't support this in the long run.

**Comment by mckn** [comment]
> I spun in circles for quite a while on how to apply CUE's native package concepts to what we're doing. Eventually, i reached the conclusion that those concepts were primarily designed with as-code repos in mind. At the level of schema/kind declaration in plugins, what we're doing just looks different. For example, there's no need for us to think about [root/medial/leaf](https://cuelang.org/docs/concepts/packages/#instances) directories in the context of a plugin, because our use pattern is not "iterate over and load a set of leaf directories, accumulating each result into a set of concrete objects." For the use case of loading kind declarations, we want to load and validate the entire set with _**exactly one**_ operation - and none of them are expected to be concrete.
> 
> Rather, we should look at what's declared in a plugin's `grafanaplugin` package as an element within that strategy - folks could plausibly `import "github.com/grafana/grafana/public/app/plugins/panel/timeseries:grafanaplugin` from a root directory in their as-code repo.

I'm new to cue as a language and have not used it extensively so I would need to trust you on this one. However, I think sticking to the way the language works and how it is structured is a good idea because:

- It will make it easier for developers with prior experience of cue to use it within the context of grafana plugin development.
- We can reference the cue lang documentation and won't need to write our own documentation.
- Leverage the existing cue lang community to be able to answer questions etc. that might occur when plugin developers start to ask questions.

**Comment by mckn** [comment]
> First, `cue.mod` is generally more like a `.git` directory - it marks the root of a the module tree, but your .cue files generally aren't supposed to go in it.
> 
> Using a tree to illustrate seems helpful, i'll do that. As i think of it, we start with literally just `plugin.json`:
> 
> ```shell
> .
> └── plugin.json
> ```
> 
> In external plugins, at least, there'll likely need to exist a `cue.mod`:
> 
> ```shell
> .
> ├── cue.mod
> │   └── module.cue
> └── plugin.json
> ```
> 
> But nothing else is in that directory.
> 
> Today, we have models.cue at the root:
> 
> ```shell
> .
> ├── cue.mod
> │   └── module.cue
> ├── models.cue # in package grafanaplugin, loaded
> └── plugin.json
> ```
> 
> And that gets loaded by logic called from `pfs.ParsePluginFS`. One shift i'm suggesting here is that we should load _any_ file sibling to `plugin.json` in the appropriate package:
> 
> ```shell
> .
> ├── cue.mod
> │   └── module.cue
> ├── grafanaplugin_file.cue # also in package grafanaplugin, loaded
> ├── models.cue
> └── plugin.json
> ```
> 
> But, to expand on this as requested
> 
> > > CUE files in any package other than grafanaplugin MAY exist, but MUST be discarded by a loader
> > > Can you expand a bit more on this?
> 
> i'm saying that it's valid for `other_pkg.cue` to exist, but `pfs.ParsePluginFS` must ignore it:
> 
> ```shell
> .
> ├── cue.mod
> │   └── module.cue
> ├── grafanaplugin_file.cue
> ├── models.cue
> ├── other_pkg.cue # in package "other", discarded
> └── plugin.json
> ```
> 
> Similarly, if there are subdirectories containing files, the loader will not automatically see them:
> 
> ```shell
> .
> ├── cue.mod
> │   └── module.cue
> ├── grafanaplugin_file.cue
> ├── models.cue
> ├── other_pkg.cue
> ├── plugin.json
> └── schemas
>     └── grafanaplugin_file.cue # not sibling to plugin.json, not loaded
> ```
> 
> In the future, we might allow `schemas/grafanaplugin_file.cue` or `other_pkg.cue` to be pulled in by `pfs.ParsePluginFS` - but only indirectly, by having an `import` statement in either `models.cue` or `grafanaplugin_file.cue` that pulls the contents of those files into the instance that `pfs.ParsePluginFS` is actually trying to load.

Ahh, mybad...I totally read the cue lang documentation. You are correct regarding the `cue.mod` folder. In my example, the schema folder should be in the root instead of in the cue.mod folder.

```bash
    .
   ├── cue.mod 
   │   └── module.cue 
   ├── schemas 
   │   ├── plugin.cue
   │   ├── entities
   │   │   ├── incident.cue
   │   │   └── ... 
   │   └── ... 
   ├── plugin.json    #we will move this to the root from the src
   ├── pkg
   ├── src
   └── ...
```

**Comment by mckn** [comment]
> In the future, we might allow `schemas/grafanaplugin_file.cue` or `other_pkg.cue` to be pulled in by `pfs.ParsePluginFS` - but only indirectly, by having an `import` statement in either `models.cue` or `grafanaplugin_file.cue` that pulls the contents of those files into the instance that `pfs.ParsePluginFS` is actually trying to load.

Yes, so this is what is was suggesting! Basically set a convention where we always try to load e.g. the `<plugin_dir>/schemas/plugin.cue`. We could allow imports but maybe limit it to be either the white listed Grafana libraries or paths within the current module.

So when parsing the plugin schemas we would build the "tree" of the plugin.cue (I guess cue already have support for this since it is available in the language) and use that as the provided schema.

**Comment by mckn** [comment]
> We should not allow this, for two reasons:
> 
> **Thema correctness** - basically, Thema's invariants can only _ever_ be meaningful if the text it's checking over for backwards compatibility, and translatability is immutable. Allowing imports of external elements is the easiest way to blow up that guarantee. Disallowing imports is neither necessary nor sufficient to address this issue, but it is a quick and easy way of doing most of what we need.
> 
> **MVP Implementation complexity** - in order to generate thema bindings (the equivalent of [this file](https://github.com/grafana/grafana/blob/49f91721ec4d76076c2de4eb2321e3039c60571a/pkg/plugins/plugindef/plugindef_bindings_gen.go), but for use in a plugin), there has to be an `embed.FS` that contains the `cue.mod` and any relevant files to load, _transitively including any imported files._ Correctly injecting the allowlisted packages is the job of functions like [this one](https://github.com/grafana/grafana/blob/49f91721ec4d76076c2de4eb2321e3039c60571a/pkg/cuectx/ctx.go#L145). But the plugin-declared bits still need to be rolled into an `fs.FS` for passing to `pfs.ParsePluginFS` - that's what's in our `embed.FS`. With the constraints i propose, this will _always_ be sufficient for that loader:
> 
> ```go
> //go:embed cue.mod *.cue
> var CueFS embed.FS
> ```
> 
> But if we allow imports, even plugin module-internal ones, we'd need more adaptive logic to decide what goes in the embed. Getting these loaders to work cleanly is deceptively challenging, and could easily delay initial delivery by weeks. Given that this is all primarily oriented towards a day 2+ scaling problem, i'm suggesting we see how far we can get with these more restrictive rules.
> 
> Now, in that `embed.FS`, the `other_file.cue` would have been included. So why disallow that in the name of MVP?

I think it would be good to be transparent about the long term goals and what is in scope for now. I think what I'm suggesting would bring a better developer experience long term. But I'm happy with keeping it out of scope for now as long as we don't end up in a situation where we won't be able to provide this in the long term.

**Comment by mckn** [comment]
> To allow imports of other sibling `.cue` files from a different package, that check ceases to be just a static lookup, and becomes dynamically dependent on state within the plugin itself. We have to know whether a `cue.mod` exists, what its mod path is, what the path relation is between `cue.mod` and `plugin.json`, whether the import statement is pointing to the same path as in that relation.
> 
> And, since `pfs.ParsePluginFS` is dependent on an `fs.FS`, not the actual filesystem, there's always the possibility that what's handed to it is a misleadingly incomplete picture of what's on the filesystem - _deeply_ confusing for users ("why is it saying i don't have a `cue.mod`? it's right there!?!?!")
> 
> These aren't unsolvable problems, and it's probably worth tackling them eventually. I just see the costs of answering them, believe we can safely and sanely defer on them for MVP, and see only marginal benefit to answering them before we're actually feeling this anticipated friction.

Okay, I think it is worth keeping in mind that once we have published something that plugin developers start to use we need to support it long term until we can deprecate it. So that is why it is important to think of the long term solution and try to scale that one down to something that is easy to deliver but won't put us in a situation where we need to do a lot of extra work to enable the more advance use case.

**Comment by sdboyer** [comment]
The theme of all these responses seems to be, "this seems OK, as long as we're not painting ourselves into a corner." And i can see how the restrictions i'm suggesting here could be seen that way.

All of these restrictions, though, are actually about not letting _users_ paint themselves, and us, into a corner: they preclude potential ambiguities and complexities that users could get lost in, and once allowed, would be very hard to take _away_ support for. But i have confidence that they're all things that can be relaxed later, once we're sufficiently confident that our tools can remain boring and reliable despite the added complexity.

**Comment by mckn** [comment]
Sounds good! But I guess we need to take some decisions on how the first version of the schema support should look towards the plugin authors.

**Comment by sdboyer** [comment]
Excellent! 

This would be the simplest possible implementation given exact state of code in main right now:

```cue
// GrafanaPluginInstance specifies what plugins may declare in .cue files within
// the `grafanaplugin` package, at the plugin root adjacent to plugin.json.
GrafanaPluginInstance: {
	// The composable kinds declared by a plugin
	composableKinds: #ComposableSlots
	// The custom kinds declared by a plugin
	customKinds: [string]: kindsys.#CustomStructured

	// it's almost certainly better to just change the way slots are declared to
	// avoid the need for this field comprehension with dynamic keys
	#ComposableSlots: {for slotname, meta in kindsys.slots {
		"\(slotname)"?: kindsys.#Composable & {
			slot: slotname
			lineage: { joinSchema: meta }
		}
	}}
}
```

**Comment by mckn** [comment]
Super! And what about file structure? Should we try to set a convention for that as well?

**Comment by sdboyer** [comment]
i think that the simplest possible one would be

```
plugin.json
composable.cue
custom.cue
```

but we could also do suffixing of the composables by their slot

```
plugin.json
composable_query.cue
composable_dsoptions.cue
custom.cue
```

**Comment by sdboyer** [comment]
Maaaaaybe? i'm assuming you're just thinking about tidiness?

i can't immediately think of a constraint that doing so would violate. I also don't know what we'd name such a directory (`kinds/`? we could replicate the declaring structure in core...), what the rationale for that name would be.

But i really don't have a strong intuitive grasp of the plugin space, and it's a sufficiently predictable pattern that i don't think it'll violate any of the invariants i originally listed.   It will be a bit more work to implement, but i don't think prohibitive. So, if it feels right to you, i'm 👍🏻

**Comment by mckn** [comment]
> Maaaaaybe? i'm assuming you're just thinking about tidiness?
> 
> i can't immediately think of a constraint that doing so would violate. I also don't know what we'd name such a directory (`kinds/`? we could replicate the declaring structure in core...), what the rationale for that name would be.
> 
> But i really don't have a strong intuitive grasp of the plugin space, and it's a sufficiently predictable pattern that i don't think it'll violate any of the invariants i originally listed. It will be a bit more work to implement, but i don't think prohibitive. So, if it feels right to you, i'm 👍🏻

I think it would be good to suggest, to the plugin authors, a simplified solution of something that they can use when the plugin code base grows. 

So my thinking is that having a number of additional `*.cue` files in the root directory of the plugin might not be optimal so setting a convention where our schema tool look in a folder for `*.cue` files might be a good idea?
